File: app/src/main/java/com/codingblocks/restapiretrofitjson/activities/AlbumsActivity.java
package com.codingblocks.restapiretrofitjson.activities;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;

import com.codingblocks.restapiretrofitjson.R;
import com.codingblocks.restapiretrofitjson.adapters.AlbumAdapter;
import com.codingblocks.restapiretrofitjson.api.AlbumsAPI;
import com.codingblocks.restapiretrofitjson.interfaces.OnItemClickListener;
import com.codingblocks.restapiretrofitjson.models.Album;

import java.util.ArrayList;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class AlbumsActivity extends AppCompatActivity {

    public static final String TAG = "ALBUMS";

    RecyclerView rvAlbumsList;
    AlbumAdapter albumAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_albums);
        rvAlbumsList = (RecyclerView) findViewById(R.id.rvAlbumsList);
        rvAlbumsList.setLayoutManager(new LinearLayoutManager(this));
        albumAdapter = new AlbumAdapter(this, new ArrayList<Album>());
        rvAlbumsList.setAdapter(albumAdapter);

        albumAdapter.setOnItemClickListener(new OnItemClickListener() {
            @Override
            public void onItemClick(int itemId,View view) {
                Intent postActIntent = new Intent(AlbumsActivity.this,PhotoActivity.class);
                postActIntent.putExtra("userId", itemId);
                startActivity(postActIntent);
            }
        });



        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl("https://jsonplaceholder.typicode.com")
                .addConverterFactory(
                        GsonConverterFactory.create()
                )
                .build();

        final AlbumsAPI albumsAPI = retrofit.create(AlbumsAPI.class);

        albumsAPI.getAlbums().enqueue(new Callback<ArrayList<Album>>() {
            @Override
            public void onResponse(Call<ArrayList<Album>> call, Response<ArrayList<Album>> response) {
                Log.d(TAG, "onResponse: ");
                albumAdapter.updateAlbums(response.body());

            }

            @Override
            public void onFailure(Call<ArrayList<Album>> call, Throwable t) {

            }
        });
    }
}

File: app/src/main/java/com/codingblocks/restapiretrofitjson/activities/AlbumsActivity.kt
package com.codingblocks.restapiretrofitjson.activities

import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.view.View
import com.codingblocks.restapiretrofitjson.R
import com.codingblocks.restapiretrofitjson.adapters.AlbumsAdapter
import com.codingblocks.restapiretrofitjson.api.API
import com.codingblocks.restapiretrofitjson.interfaces.OnItemClickListener
import com.codingblocks.restapiretrofitjson.models.Album

import kotlinx.android.synthetic.main.activity_albums.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*

class AlbumsActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_albums)

        rvAlbumsView.layoutManager = LinearLayoutManager(this)
        val albumsAdapter = AlbumsAdapter(ArrayList<Album>(), this)
        rvAlbumsView.adapter = albumsAdapter
        pbProgress.visibility = View.VISIBLE
        btnShowAllPhotos.setOnClickListener { startActivity(Intent(this@AlbumsActivity, PicsActivity::class.java)) }
        albumsAdapter.setOnItemClickListener(OnItemClickListener { itemid, view ->
            val i = Intent(this@AlbumsActivity, PicsActivity::class.java)
            i.putExtra("albumId", itemid)
            startActivity(i)
        })
        val receivedId = intent.getIntExtra("userId", -1)
        supportActionBar!!.title = "Albums"
        val albumsApi = API.getInstance().albumsAPI
        val receiveAlbums = object : Callback<ArrayList<Album>> {
            override fun onResponse(call: Call<ArrayList<Album>>, response: Response<ArrayList<Album>>) {
                albumsAdapter.updateAlbums(response.body()!!, pbProgress)
            }

            override fun onFailure(call: Call<ArrayList<Album>>, t: Throwable) {

            }
        }

        if (receivedId != -1) {
            albumsApi.getAlbumsbyUserId(receivedId).enqueue(receiveAlbums)
        } else {
            albumsApi.getAlbumsbyUserId(null).enqueue(receiveAlbums)
        }

    }
}